[DataSpace] FR 45527: Add link to static blogs on public page

labkey-bpatel opened the pull request:
#### Rationale
Add link to static blogs on public page. Other minor updates to the public page.
Spec [here](https://docs.google.com/document/d/1aYgnxAsJRPW8kTXnif8aQoU1gmaGsWdH2OeZ_qZIcQI/edit)

#### Changes
* Updates to get blog data on public page. 
* Add module properties for blog directory and all blogs.
* Update FrontPage to add new blog section and display four recent blogs.
* Convert images to thumbnail.
* Updates to the Welcome section to display the full 'down arrow' button.
* Updates to Always Growing section to use a pill background for stats.
* Styling updates.
* Add sample data.

labkey-bpatel committed:
Updates to get blog data on public page. Update FrontPage to display four recent blogs (wip).

labkey-bpatel committed:
Generate thumbnails from an image. Add sample images.

labkey-bpatel committed:
Styling updates

labkey-bpatel committed:
Add a section with 'Recent blog post' and 'Go to blog' button

labkey-bpatel committed:
More styling updates. Move inline styling to .scss.

labkey-bpatel committed:
More styling updates, adjust spacing/padding/margin

labkey-bpatel committed:
Updates to the Welcome section to display the full down button.

labkey-bpatel committed:
Move Blogs to its own section on the public page

labkey-bpatel committed:
Update sample data

labkey-bpatel committed:
Cleanup

labkey-bpatel committed:
Sample data updates, use module properties

labkey-bpatel committed:
Cleanup

labkey-klum committed:
Merge branch 'develop' into fb_static_blog_public_page

labkey-klum commented on src/org/labkey/cds/CDSController.java:
```diff
@@ -191,12 +194,21 @@ public Object execute(Object o, BindException errors) throws Exception
+
+            // In order to display blog posts on Dataspace's public page where the user is not required to log-in, we need to create a
+            // Limited user with Reader role otherwise no feeds will be returned by RSSService.get().getFeeds()
+            if (getUser().isGuest())
```
I don't think we care if the user is a guest or not but rather if the user had read access to the folder:

```suggestion
            if (getContainer().hasPermission(user, ReadPermission.class))
```

labkey-bpatel replied:
getContainer().hasPermission(user, ReadPermission.class) evaluates to false. Guest user doesn't have a read permission, hence the reason to create a LimitedUser with ReaderRole.

labkey-klum replied:
sorry missed a `!` in front of that conditional it should have read more like:
```
if (!getContainer().hasPermission(user, ReadPermission.class))
```
Unless we are really trying to distinguish between guest and non-guest users, checking the access level is the preferred route.

labkey-bpatel replied:
Ah, got it! That makes sense, thank you for clarifying. Now updated!

labkey-klum commented on src/org/labkey/cds/CDSController.java:
```diff
@@ -191,12 +194,21 @@ public Object execute(Object o, BindException errors) throws Exception
+            // Limited user with Reader role otherwise no feeds will be returned by RSSService.get().getFeeds()
+            if (getUser().isGuest())
+            {
+                user = new LimitedUser(User.guest, new int[0], Collections.singleton(RoleManager.getRole(ReaderRole.class)), false);
```
We should also be able to reuse the current user object versus `User.guest`
```suggestion
                user = new LimitedUser(user, new int[0], Collections.singleton(RoleManager.getRole(ReaderRole.class)), false);
```

labkey-bpatel replied:
Good point. Done!

labkey-klum commented on src/org/labkey/cds/CDSManager.java:
```diff
@@ -330,6 +329,18 @@ public String getCDSImportFolderPath(Container container)
+    public String getBlogPath(Container container)
+    {
+        ModuleProperty mp = ModuleLoader.getInstance().getModule(CDSModule.class).getModuleProperties().get(CDSModule.BLOG_PATH);
+        return PageFlowUtil.decode(mp.getEffectiveValue(container));
```
It would be nice if this method could tolerate either a relative or fully qualified URL but then return a fully qualified URL. That way the code in `frontPage.jsp` wouldn't need to concatenate prepend `baseURL` to the `blogPath`. For example the module property could be either:
- `/_webdav/home/Public/%40files/publicPage/blogs/`
- `http://localhost:8080/labkey/_webdav/home/Public/%40files/publicPage/blogs/`

but always include the base URL if necessary.

labkey-bpatel replied:
Done

labkey-klum commented on src/org/labkey/cds/view/template/FrontPage.jsp:
```diff
@@ -73,6 +80,69 @@
+                            document.getElementById("thumbnail" + (imageIdx+1)).appendChild(canvas);
+                        };
+
+                        $('<td>').html('<a id="thumbnail1" href="' + items[0].link + '"></a>').appendTo($tr);
```
Could we have a value for the max number of thumbnails to render and just loop over the news feed items to create the elements? It looks like this code will blow up if less than 4 items are returned from the news feed. 

Do we need to be more fault tolerant on whether all item properties may be present? Or is this validation happening elsewhere?

labkey-bpatel replied:
Now iterating through max number of items.

Did not make it fault tolerant (or initially add any checks) since these blog items are already in use to display on the Dataspace's home page, and there already are more than 4 blogs present, and new ones will always get added. And if the image file is not found or the thumbnail property is absent in the json file then it will show an empty box like so:
![image](https://github.com/LabKey/cds/assets/10765266/dfe935db-3e61-46e7-ab13-ff8dedc23ce2)

labkey-klum commented on src/org/labkey/cds/view/template/FrontPage.jsp:
```diff
@@ -73,6 +80,69 @@
+                            var img = new Image();
+                            img.src = "<%=h(baseURL + blogPath)%>" + items[imageIdx].imageFile;
+
+                            console.log("img src = " + img.src);
```
left over debug code?

labkey-bpatel replied:
Cleared

labkey-klum commented on theme/front-page/index/sections.scss:
```diff
@@ -20,7 +20,7 @@
 
   .fp-scrollable {
-    padding-top: 6em !important;
+    //padding-top: 6em !important;
```
should we delete this?

labkey-bpatel replied:
Deleted

labkey-klum commented on theme/front-page/index/sections.scss:
```diff
@@ -69,11 +69,11 @@
-    @include legacy-desktop {
-      margin: 10.67em 0 2.67em 0;
-    }
+    //@include legacy-desktop {
```
is this needed any longer?

labkey-bpatel replied:
Removed

labkey-klum reviewed (changes requested):
see comments with suggested changes.

labkey-bpatel committed:
Code review changes:
- Reuse current user
- Tolerate either a relative or fully qualified URL for blog path module property
- Use max number of new items to display
- Cleanup

labkey-bpatel committed:
Iterate through the max items to display.

labkey-bpatel committed:
Client requested changes:
- Use 'News' instead of 'Blog Posts'
- Change color on 'Recent Blog Posts' text

labkey-bpatel committed:
Replace 'imageFile' property name with 'thumbnail'.

labkey-bpatel committed:
Code review change: check read permission

labkey-bpatel committed:
Merge remote-tracking branch 'origin/develop' into fb_static_blog_public_page

labkey-bpatel committed:
Update test to match the updated sample data
